### ISSUE TITLE
Structure of new "meta" objects

### ISSUE THREAD

**Issue by Nate-Fulbright** [AlexDenman/mtreetools#1]
@AlexDenman, could you take a look through the new "mtreeMeta" folder I added in the branch "metaNate"?

Here's the high level structure:
![image](https://github.com/AlexDenman/mtreetools/assets/159860517/d727e291-7a1a-447b-b50d-40766a3f9f9f)

What are your thoughts on the overall organization?

I took the approach of abstracting more of the declared variable functionality vs. how you approached it in your demo of the "Argument" class. I did this so the object itself makes it more clear what its purpose is rather than needing to look in at the specific properties. I know you said you weren't normally a fan of inheritance, but I think this is a good application for it.

Before either of us put much more work into setting this up, let's try to agree on a structure and property names within that. We can start with the "Argument" object since that's the one you've uploaded so far.

Overall, the properties we have are similar. Here's a summary of your properties:
![image](https://github.com/AlexDenman/mtreetools/assets/159860517/5807855b-845b-4101-a8fd-40d630f33e76)

Here is what I have:
![image](https://github.com/AlexDenman/mtreetools/assets/159860517/2421ebea-66a1-4319-8a7b-3a6cf827a38c)

My specific proposals:

- Use lower camel case for anything that's not an object or structure. `name` instead of `Name`, `isRepeating` instead of `IsRepeating`, etc.
- Make `ParentFunction` into a more generic `Parent` that has type of `mtreeMeta.Function`. This would allow for both function parents and method parents.
- Use `classValidation` and `sizeValidation` instead of `Class` and `Size` to better represent that the class and size by the argument are validation and not "properties" of the property.
- Use `validationFunctions` instead of `ValidatorFunctions` to better align with MATLAB terminology ([source](https://www.mathworks.com/help/matlab/matlab_prog/argument-validation-functions.html)).
- Use the various implementations of the `DeclaredVariable` class to represent the information like `IsOutput` and `IsProperty`. This information would be conveyed by the fact that the object is `mtreeMeta.Property` or `mtreeMeta.OutputArgument`.
- Replace `Keyword` with `nameValueStructureName` to align with [MATLAB seeming to refer to the structures as "name-value" structures](https://www.mathworks.com/help/matlab/matlab_prog/validate-name-value-arguments.html#:~:text=when%20using%20multiple-,name%2Dvalue%20structures,-Cannot%20use%20names). I'll rename the property in the `InputArgument` class if we take this route.

Lots for you to look through here, take your time to process and respond.

Once we get this figured out, let's discuss how to best split up the remaining tasks.

**Comment by AlexDenman** [comment]
Most of this looks good to me. I'm fine with organizing it into a class hierarchy this way. The only part I'm unsure about is that some of these classes seem like they might unnecessarily duplicate functionality already available in matlab's "meta" package. In particular, the "Class" class and the "Property" class seem like they duplicate meta.class and meta.property. What is the reason for including those classes?

For functions, of course, we need a new class, because no corresponding one exists. But for e.g. the Property class, the properties listed here seem like a carbon copy of meta.property, so why not just create and use meta.property objects instead? 

Also, for DeclaredVariable, now that I think about it I'm not sure the names ClassValidation and SizeValidation make sense, because unlike the validation functions, those two properties don't just validate the input, they alter it. I think this is why the corresponding properties of meta.Validation are named Class and Size. What do you think of declaredClass and declaredSize, or enforcedClass / enforcedSize?

**Comment by AlexDenman** [comment]
Some additional comments after looking at some of the files:

What is the reason for making the class properties not immutable / defining setter methods for some of the properties? As far as I can tell, all the properties of the various metadata objects seem like their proper values should be known at the time of their creation, and thus I would think it makes sense to make those properties immutable. I'm curious what the reason for doing it this way is.

**Comment by Nate-Fulbright** [comment]
> Most of this looks good to me. I'm fine with organizing it into a class hierarchy this way. The only part I'm unsure about is that some of these classes seem like they might unnecessarily duplicate functionality already available in matlab's "meta" package. In particular, the "Class" class and the "Property" class seem like they duplicate meta.class and meta.property. What is the reason for including those classes?

For `Property`, the main reason would be consistency with the rest of our `DeclaredVariable` objects. That way any post-processing (for things like headers) wouldn't need to have special logic to handle the MATLAB `meta` objects. Similarly for `Class`, this would point to our `Method` and `Property` objects instead of the `meta.method` and `meta.property` objects. The backend would basically just be processing the `meta` objects.

> Also, for DeclaredVariable, now that I think about it I'm not sure the names ClassValidation and SizeValidation make sense, because unlike the validation functions, those two properties don't just validate the input, they alter it. I think this is why the corresponding properties of meta.Validation are named Class and Size. What do you think of declaredClass and declaredSize, or enforcedClass / enforcedSize?

That's a good point. And seeing your proposals `declaredClass` or `enforcedClass`, that makes me realize just `size` and `class` are probably descriptive enough, and there really isn't a practical difference between "this is the size that will be validated" and "this is the size that the object will be enforce/converted to".

So then, assuming we use `size` and `class`, do we keep the corresponding `hasSizeValidation` and `hasClassValidation`? Can you think of a better name? Of course `size` and `class` will just have a default value of `string.empty`, but I've still found it helpful to have the explicit `hasXXX` variables too. `hasSize` and `hasClass` just feels off to me, but maybe I can be convinced.

> What is the reason for making the class properties not immutable / defining setter methods for some of the properties? As far as I can tell, all the properties of the various metadata objects seem like their proper values should be known at the time of their creation, and thus I would think it makes sense to make those properties immutable. I'm curious what the reason for doing it this way is.

Good question! I actually did originally have all of the properties as `SetAccess = immuatable` originally, for exactly the reason you mentioned. Thinking through it some more as well, maybe it does actually make sense to change them back to `SetAccess = immuatable`. When we create these, we'd be doing something like `mtreeMeta.Property(somePropertyDataFromMtree)` and the constructor would do the processing it needs and set the variables. With `SetAccess = private`, we get the ability to do `P = mtreeMeta.Property; P.processMtree(somePropertyDataFromMtree);`, but in the context of this tool where we don't have any other publicly settable properties, this isn't as useful.

To summarize that rambling, I agree, let's change to `SetAccess = immuatable`. I'll still leave the set methods to do the validation to help keep things tidier.

I have 2 hours of meetings coming up right now, but then I can make these changes. I'll only do the `SetAccess = immuatable` stuff for now until I hear back from you for your thoughts on the `Property`, `Class`, and `size`/`class` stuff.

**Comment by AlexDenman** [comment]
> For Property, the main reason would be consistency with the rest of our DeclaredVariable objects. That way any post-processing (for things like headers) wouldn't need to have special logic to handle the MATLAB meta objects. Similarly for Class, this would point to our Method and Property objects instead of the meta.method and meta.property objects. The backend would basically just be processing the meta objects.

This seems like it adds a lot of complexity. It also means more work in the future as new properties get added to the `meta` classes. Also, unless we REALLY go nuts on reimplementing features from `meta`, we lose out on some potentially useful functionality. For example, `meta.class` has special `==` / `~=` / `>` / `>=` / etc. methods for testing class equality, subclass-ness, and so on. 

Here is what I would propose as an alternative:

1. We remove the `Method` class and give our `Function` objects the properties `classMethod` and `packageFunction` (both of class `meta.method`). If the function is a class method, then we instantiate the appropriate `meta.method` object and store it in the `classMethod` property, and likewise for `packageFunction` if it's a package function. Additionally we add "package function" as a possible value for the `type` property of the `Function` class.
2. We remove the `Property` class and give our `DeclaredVariable` class a `classProperty` property, of type `meta.property`.
3. We give the `Argument` superclass a `parentFunction` property, of type `Function`.
4. We remove the `Class` class, because one can obtain the parent `meta.class` object for any object that is contained in a class by traversing the object references. For example, if `foo` is a `DeclaredVariable` that represents an input argument to a class method, we take `foo.parentFunction.classMethod.DefiningClass`, which will return the associated `meta.class` object. In my opinion, this should emulate the behavior of `meta` objects, which is that if the object does not HAVE e.g. a `DefiningClass` then that property is empty, and the result of the above expression would return an empty `meta.class` rather than erroring.
5. We add some type of lookup functionality to enable finding/instantiating the appropriate `Function` object for a given `meta.method`, or the appropriate `DeclaredVariable` object for a given `meta.property`. This is necessary because we can't directly extend `meta` classes to make those classes store `Function` or `DeclaredVariable` objects to enable traversing downwards, from a class / package down to a function / argument. (Upwards traversal is already possible because we are storing `meta` objects as properties of our objects.) This maximizes the potential for code that already uses `meta` for introspection to use our objects.

I'm not entirely happy with my suggestion #2. Two alternatives are:
A: Keeping `Property` but having it simply define one extra property, which holds the relevant `meta.property` object
B: Making `DeclaredVariable` a property of `Argument` rather than `Argument` being a subclass of `DeclaredVariable`
- `Argument` inherits from `Metadata` instead of `DeclaredVariable`
- `DeclaredVariable` gets properties `parentArgument` (type `Argument`) and `parentClassProperty` (type `meta.property`)
- `Argument` gets property `VariableDeclaration` (type `DeclaredVariable`)

Option B is the most radical change, but it appeals to me because it reflects the relationship between function arguments and variable declarations: variable declarations can belong to an argument or class property, but both arguments and class properties may not have a variable declaration. E.g. for a function with no argument blocks, we would create `Argument` objects but no associated `DeclaredVariable`. Lastly (and this may not be all that relevant), option B also much more closely reflects the underlying syntactic relationships between properties/arguments and variable declarations.

Optionally, we could use `matlab.mixin.indexing.RedefinesDot` to check if the requested property is a property of the parent object and return it if it is, so you can directly obtain properties of the parent `meta` object without the extra indexing step.

Overall, the benefits I'm aiming to achieve with the above suggestions are:
- Minimized duplication of function
- Easier maintenance, by minimizing the need for updates to account for change in Matlab's `meta` classes
- Better compatibility across versions (if `meta` features are added, any code we write that depends on them would break on older versions)
- Fuller access to methods of the existing `meta` classes
- Reduced complexity on our end (fewer classes that need implementation)
- Better compatibility of our code with existing code that makes use of `meta` objects
- Single Source of Truth: By removing properties that are duplicated across classes, and enforcing that `meta` objects are instantiated at the same time as our objects, we make it harder for our code to ever give a different answer about e.g. a class method's `Access` from the answer given by `meta.method`.

What do you think of the above?

**Comment by AlexDenman** [comment]
Also, since it may help illustrate what I'm envisioning, here is a crude entity relation diagram of what I am proposing (specifically, option B for how to handle `DeclaredVariable` / `Property`):

![meta_references](https://github.com/AlexDenman/mtreetools/assets/47120300/263069e8-4f7d-45f0-9ac8-19de7ff26566)

For simplicity, the Argument subtypes are shown as a single entity, and the diagram omits the link between class properties and their set/get methods.

Red objects are Matlab's `meta` classes, and blue are our new `Metadata` objects.

Note: Arrows do NOT represent inheritance / subclassing. Green bi-directional arrows represent relationships where objects of both classes store references to each other, e.g. a `meta.class` object stores an array of `meta.property` objects, and each `meta.property` object stores the parent `meta.class` object. I like this pattern, and I think we should emulate it wherever possible. The only area where this is impossible is that we can't make `meta.method` objects reference `Function` objects, nor `meta.property` objects reference `DeclaredVariable` objects (unidirectional yellow arrows), hence the need for some kind of reverse lookup (possibly implemented by letting our class constructors accept `meta` objects of the appropriate type as inputs).

**Comment by Nate-Fulbright** [comment]
I've read through your responses and am working on thinking through them and replying to them. I've got a busy afternoon, so I probably won't get a response in until tonight.
